src/repository.rs

```diff
@@ -117,136 +125,155 @@ impl<D> Repository<D> for FileSystemRepository<D>
-            File::open(&path)?,
-            max_size.unwrap_or(::std::usize::MAX) as u64,
-            min_bytes_per_second,
-            hash_data,
-        )?;
+                let mut reader = SafeReader::new(
+                    AllowStdIo::new(File::open(&path)?),
+                    max_size.unwrap_or(::std::usize::MAX) as u64,
+                    min_bytes_per_second,
+                    hash_data,
+                )?;
 
-        Ok(D::from_reader(read)?)
+                let mut buf = Vec::with_capacity(max_size.unwrap_or(0));
+                await!(reader.read_to_end(&mut buf))?;
+
+                Ok(D::from_reader(buf.as_slice())?)
+            },
+        )
     }
 
-    fn store_target<R>(&self, mut read: R, target_path: &TargetPath) -> Result<()>
+    fn store_target<'a, R>(
+        &'a self,
+        mut read: R,
+        target_path: &'a TargetPath,
+    ) -> TufFuture<'a, Result<()>>
     where
-        R: Read,
+        R: AsyncRead + 'a,
     {
-        let mut path = self.local_path.join("targets");
-        path.extend(target_path.components());
+        Box::pinned(
+            async move {
+                let mut path = self.local_path.join("targets");
+                path.extend(target_path.components());
 
-        if path.exists() {
-            debug!("Target path exists. Overwriting: {:?}", path);
-        }
+                if path.exists() {
+                    debug!("Target path exists. Overwriting: {:?}", path);
+                }
 
-        atomically_write(&path, |write| {
-            io::copy(&mut read, write)?;
-            Ok(())
-        })
+                let mut temp_file = AllowStdIo::new(create_temp_file(&path)?);
+                await!(read.copy_into(&mut temp_file))?;
+                temp_file.into_inner().persist(&path)?;
+
+                Ok(())
+            },
+        )
     }
 
-    fn fetch_target(
-        &self,
-        target_path: &TargetPath,
-        target_description: &TargetDescription,
+    fn fetch_target<'a>(
+        &'a self,
+        target_path: &'a TargetPath,
+        target_description: &'a TargetDescription,
         min_bytes_per_second: u32,
-    ) -> Result<SafeReader<Self::TargetRead>> {
-        let mut path = self.local_path.join("targets");
-        path.extend(target_path.components());
+    ) -> TufFuture<'a, Result<Box<dyn AsyncRead>>> {
+        Box::pinned(
+            async move {
+                let mut path = self.local_path.join("targets");
+                path.extend(target_path.components());
 
-        if !path.exists() {
-            return Err(Error::NotFound);
-        }
+                if !path.exists() {
+                    return Err(Error::NotFound);
+                }
 
-        let (alg, value) = crypto::hash_preference(target_description.hashes())?;
+                let (alg, value) = crypto::hash_preference(target_description.hashes())?;
+
+                let reader: Box<dyn AsyncRead> = Box::new(SafeReader::new(
```

Comment:
Why change this from `SafeReader` directly to `Box`ing it?

Reply:
Do you mean changing this into:

```rust
let reader = SafeReader::new(...);
Ok(Box::new(reader) as Box<dyn AsyncRead>)
```

?, if so, yeah I could switch to that. Or are you referring to hiding the fact that we're using an `AsyncRead`? If the latter, I'm a little back and forth on it. The `FilesystemRepository` doesn't really need the full `SafeReader` interface, it really just needs a reader that asserts the hash is correct. That suggested to me that we might want to break up the `SafeReader`, assuming there's a performance benefit. But we lose guaranteeing that other impls of `Repository` have to protect against the things `SafeReader` protects against. I go back and forth.

@heartsucker: do you have an opinion on this?

Reply:
This was originally done for simplicity and for being able to reuse the `Repository` type as both local and remote. It should probably be two types because local repositories can be semi-trusted. Whatever gets this merged in is fine by me.